homeassistant/components/tplink/config_flow.py

```diff
@@ -52,15 +78,88 @@ async def _async_handle_discovery(self, host: str, mac: str) -> FlowResult:
         for progress in self._async_in_progress():
             if progress.get("context", {}).get(CONF_HOST) == host:
                 return self.async_abort(reason="already_in_progress")
-
+        credentials = await get_stored_credentials(self.hass)
         try:
-            self._discovered_device = await self._async_try_connect(
-                host, raise_on_progress=True
+            await self._async_try_discover_connect(
+                host, credentials, raise_on_progress=True
             )
+        except AuthenticationException:
+            self._last_failed_discovery_credentials = credentials
+            return await self.async_step_discovery_auth_confirm()
         except SmartDeviceException:
             return self.async_abort(reason="cannot_connect")
+
         return await self.async_step_discovery_confirm()
 
+    async def async_step_discovery_auth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that auth is required."""
+        assert self._discovered_device is not None
+        errors = {}
+        host = self.context[CONF_HOST]
+        credentials = await get_stored_credentials(self.hass)
+
+        if credentials != self._last_failed_discovery_credentials:
+            httpx_asyncclient = create_async_httpx_client(self.hass, verify_ssl=False)
+            try:
+                device = await Discover.connect_single(
+                    host,
+                    port=self._discovered_device.port,
+                    credentials=credentials,
+                    connection_params=self._discovered_device.connection_parameters,
+                    httpx_asyncclient=httpx_asyncclient,
+                    try_discovery_on_error=False,
+                )
+            except AuthenticationException:
+                pass
```

Comment:
This should likely be handled somehow?

Reply:
There's already an auth error so this part of the step is just trying again if the credentials have changed since the error and it will skip the UI if it works.  If it doesn't work then it moves on and shows the UI.  Make sense?